WISE-Community/WISE-API
# feat(Get Classmate Work): Create endpoint for classmate Summary work

- Make sure the student can retrieve data from the other step that is referenced in the Summary component.

- Make sure the student can not retrieve data from another step if that other step is not referenced in the Summary component.
  - For example there could be a legitimate
summaryNode1, summaryComponent1, legitOtherNode, legitOtherComponent
but the student tries to retrieve
summaryNode1, summaryComponent1, notLegitOtherNode, notLegitOtherComponent

Example Summary student work endpoint for period
```
http://localhost:81/api/classmate/summary/student-work/RUN_ID/PERIOD_ID/node292/if1755lrin/node231/0k0r4jlona/period
```

Example Summary student work endpoint for all periods in run
```
http://localhost:81/api/classmate/summary/student-work/RUN_ID/PERIOD_ID/node292/if1755lrin/node231/0k0r4jlona/allPeriods
```

Example Summary annotations endpoint for period
```
http://localhost:81/api/classmate/summary/annotations/RUN_ID/PERIOD_ID/node292/if1755lrin/node231/0k0r4jlona/period
```

Example Summary annotations endpoint for all periods in run
```
http://localhost:81/api/classmate/summary/annotations/RUN_ID/PERIOD_ID/node292/if1755lrin/node231/0k0r4jlona/allPeriods
```

- Test that project content caching and cache evicting works.

You can test the caching by modifying the ProjectServiceImpl.java getProjectContent() function to add the try/catch that sleeps for 3 seconds.
```
  @Cacheable(value = "projectContent", key = "#project.getId()")
  public String getProjectContent(Project project) throws IOException {
    try {
      long time = 3000L;
      Thread.sleep(time);
    } catch (InterruptedException e) {
      throw new IllegalStateException(e);
    }
    String projectFilePath = appProperties.getProperty("curriculum_base_dir")
        + project.getModulePath();
    return FileUtils.readFileToString(new File(projectFilePath));
  }
```


You can test the automatic cache evicting after a certain amount of time by changing the Timer.ONE_DAY to Timer.ONE_MINUTE.
```
  @Scheduled(fixedRate = Timer.ONE_MINUTE)
  @CacheEvict(value = "projectContent", allEntries = true)
  public void evictAllProjectContentCache() {
  }
```

How to test caching
1. Log in as a teacher in one browser
2. Log in as a student in another browser
3. Have the student launch a run that contains a Discussion component
4. Have the teacher load the Authoring tool for that project run
5. In the student browser, manually visit the url to retrieve the classmate Discussion student work. This should take 3 seconds if you added the sleep code above.
```http://localhost:81/api/classmate/discussion/student-work/RUN_ID/PERIOD_ID/NODE_ID/COMPONENT_ID```
6. Refresh the url and it should load instantly (instead of taking 3 seconds)
```http://localhost:81/api/classmate/discussion/student-work/RUN_ID/PERIOD_ID/NODE_ID/COMPONENT_ID```
7. Wait 1 minute for the auto cache evict to run
8. Refresh the url and it should take 3 seconds to load
```http://localhost:81/api/classmate/discussion/student-work/RUN_ID/PERIOD_ID/NODE_ID/COMPONENT_ID```
10. Go to the teacher browser and make a change to the project in the Authoring Tool.
11. Go back to the student browser and refresh the url. It should take 3 seconds to load.
```http://localhost:81/api/classmate/discussion/student-work/RUN_ID/PERIOD_ID/NODE_ID/COMPONENT_ID```

Closes #55
